### ISSUE TITLE
Segment Violation when using fullhook.so with FireFox on Ubuntu 16.04 (xenial)

### ISSUE THREAD

**Issue by rsundahl** [polyverse/readhook#2]
Fullhook.so and basehook.so work as described on Ubuntu xenial with applications socat and vi, but not with FireFox.

**Comment by rsundahl** [comment]
I suspected that this was caused by the large number of threads that FireFox starts, but the cause is in the following line:
```
char *p = (result < strlen(s_fullmagic)) ? NULL : strnstr(buf, s_fullmagic, result);
```
and more specifically in the comparison:
```
(result < strlen(s_fullmagic))
```
This is because result is of type ssize_t, which is signed, but strlen() is of type size_t which is unsigned (at least in this case). I am surprised that this hasn't bitten me more (no warnings about this and I'm sure I've done it before). When result is -1, the comparison fails because it is compared to strlen() as a large unsigned value. The correct way to compare size_t and ssize_t then is:
```
(result < (ssize_t) strlen(s_fullmagic))
```
The change was made in src/fullhook.c and src/halfhook.c and will be included in v1.2.2.

**Comment by rsundahl** [comment]
Closed with release of v1.2.2. (Note: While readhook is now playing well with firefox, upon the actual overflow, a reverse-shell is nevertheless not created. The attempt appears to be trapped by the firefox sandboxes but isn't yet clear. A new issue for further exploration will be created.)
